ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java

```diff
@@ -287,16 +302,27 @@ private void buildAddClauseForAlter(StringBuilder query) {
   private void buildSelectClauseForInsert(StringBuilder query) {
     // Need list of columns for major crud, mmmajor partitioned, mmminor
     List<FieldSchema> cols;
-    if (major && crud || major && insertOnly && sourcePartition != null || minor && insertOnly) {
+    if (rebalance || major && crud || major && insertOnly && sourcePartition != null || minor && insertOnly) {
       if (sourceTab == null) {
         return; // avoid NPEs, don't throw an exception but skip this part of the query
       }
       cols = sourceTab.getSd().getCols();
     } else {
       cols = null;
     }
-
-    if (crud) {
+    if (rebalance) {
+      query.append("0, t2.writeId, t2.rowId % ");
+      query.append(numberOfBuckets);
```

Comment:
Is the logic to assign bucketId = rowId % numberOfBuckets?

Reply:
Yes, in order to equally distribute the records among the buckets, the numberOfBuckets % row_number is used.

Reply:
Understood. Thanks for the info.